libyul/optimiser/UnusedFunctionArgumentPruner.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * UnusedFunctionArgumentPruner: Optimiser step that removes unused-parameters function.
+ */
+
+#include <libyul/optimiser/UnusedFunctionArgumentPruner.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+
+namespace
+{
+
+/**
+ * First step of UnusedFunctionArgumentPruner: Find functions with whose parameters are not used in
+ * its body.
+ */
+struct FindFunctionsWithUnusedParameters
+{
+	void operator()(FunctionDefinition const& _function)
+	{
+		auto namesFound = ReferencesCounter::countReferences(_function.body);
+
+		TypedNameList reducedParameters;
+
+		for (auto const& parameter: _function.parameters)
+			if (namesFound.count(parameter.name))
+				reducedParameters.push_back(parameter);
+
+		if (reducedParameters.size() < _function.parameters.size())
```

Comment:
If we run this step again without the function being inlined, we will add more and more functions, so maybe add an exception here:
Do not apply the step to the function if
 - the body is empty or
 - it only consists of a single ExpressionStatement with a single function call
 - it only consists of a single assignment with a single function call

Maybe we could also just reduce it to: do not apply if the number of statements in the body is at most 1.